util/exec_binary.go

```diff
@@ -0,0 +1,46 @@
+package util
```

Comment:
Again, let's try to stay away from executing binaries on the server, if we can help it.
https://github.com/giorgisio/goav
https://github.com/anacrolix/ffprobe
https://github.com/vansante/go-ffprobe

Reply:
These libraries are actually wrappers for ffprobe binary and they execute it underneath, and it think it's better to execute it yourself because:
1. You construct a command that only extracts what it needs not whole information about the video, which _may_ be slower.
2. No need to relay on repository which may be deleted.
3. No need for digging into library in case of some problems or if you need to make things differently.
I've used the go-ffprobe but then the problem of IP leakage appeared and first thought was to use proxy what is supported by ffporbe but not the wrapper that's why I had to execute it directly. (Proxy idea was eventually replaced by whitelisting domains, since it would hide server's IP but not clients' IPs.)

Reply:
So it seems, that's a shame.
https://github.com/anacrolix/ffprobe/blob/95e92c7ca8060056c59c116e87ae5a63e4750152/init.go#L19
>You construct a command that only extracts what it needs not whole information about the video, which may be slower

I'm not experienced with ffprobe, but I don't see any way it could extract information without downloading and parsing the entire source, so unless I'm wrong, this is premature optimization and reinventing the wheel.

>No need to relay on repository which may be deleted.
No need for digging into library in case of some problems or if you need to make things differently.

Both seem stable to me, I don't think it justifies the extra complexity over using the library.
Wouldn't IP leakage be a problem either way with either ffprobe or go-ffprobe? We're talking about server -> requested source, right?
We're already downloading the video and adding it to the server's filesystem with this, so no IP leaks should be propagated to the client(s), unless we're also sending unfiltered metadata/url and such.

Reply:
Also seems like https://github.com/iawia002/annie has at least basic video information, unless we need more than time/title/size/etc?

Reply:
>I'm not experienced with ffprobe, but I don't see any way it could extract information without downloading and parsing the entire source, so unless I'm wrong, this is premature optimization and reinventing the wheel.

It's not downloaded, I believe that duration is stored in metadata at the beginning of file.
> Wouldn't IP leakage be a problem either way with either ffprobe or go-ffprobe? We're talking about server -> requested source, right?

I said that as an example where having own wrapper is better, because it required modification of the command, which is not possible with wrapper.
>We're already downloading the video and adding it to the server's filesystem with this, so no IP leaks should be propagated to the client(s), unless we're also sending unfiltered metadata/url and such.

Videos are not not distributed by the server, it could take long to download them, clients just receive source URL.

Reply:
Source URL will lead to IP leaks, specifically with youtube videos.
There's a reason we had my JS-less youtube implementation removed.

Reply:
We should use noembed for these URLs, or download them to prevent URL IP leaks.

Reply:
Right, I nearly forgot that even with annie and youtube-dl we'd risk IP leaks too with certain websites, since it fetches the source URL and not the video page.

Reply:
Annie does however, support an HTTP and a SOCKS proxy, I'll find out if we can get metadata without (real) IP leaks.

Reply:
I think we may be able to, however the proxy used has to be a locally configured proxy, it's not a proxy that accepts `ip`, `port`, `user`, `password`, that will require extra setup, we'll have to see what lat thinks.

Reply:
I don't think proxy is needed.
Youtube and invidious are trustworthy and later DDoS the server
and raw sources must be added to whitelist by server admin.

Reply:
Proxy is needed when using youtube-dl functionality, googlevideo url contains some leaks, and cannot work without said leaks, thus the proxy.
This may or may not apply to other video services.

Reply:
Said proxy with only annie would also eliminate trusting the video provider altogether, along with the aforementioned.
Again, let's get lat's input before proceeding further, because if this can work, I'll be bringing back my JS-less youtube stuff, maybe use it for bitchute, and your PR would then hook into that for some parts, since common code.

Reply:
>Proxy is needed when using youtube-dl functionality, googlevideo url contains some leaks, and cannot work without said leaks, thus the proxy.
I don't understand, can you elaborate on that?

Reply:
youtube-dl grabs googlevideo url source to download from, googlevideo url contains IP leak.

Reply:
So, sending googlevideo url source to client would result in leak propagation.
Same thing happened with my JS-less youtube thing before I found out semi-recently.

Reply:
Anyway, doesn't work by removing the leak, tried like 10 different ways, so gotta do different IP instead for this to work, as well as my own youtube stuff to work.
Thanks to you I found annie.

Reply:
>So, sending googlevideo url source to client would result in leak propagation.

But that is not how it's done, client gets the youtube video, not extracted video source.
You should try running this change :D

Reply:
Then why is youtube-dl being installed on the server?

Reply:
> Again, let's get lat's input before proceeding further, because if this can work, I'll be bringing back my JS-less youtube stuff, maybe use it for bitchute, and your PR would then hook into that for some parts, since common code.

That would be nice
youtube-dl also supports proxy

Reply:
Ok but only server sees it

> Then why is youtube-dl being installed on the server?

Only to get video duration, server must know when to pop video from playlist.

Reply:
and also title.

Reply:
so... this is... the power of... speedreading...
hahaha...
![1423583808712](https://user-images.githubusercontent.com/30435868/65641303-099d7e80-dfb2-11e9-913b-10703ad79fda.jpg)